// element-plus/lib/components/config-provider/src/config-provider.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var vue = require('vue');
var props = require('../../../utils/props.js');
require('../../../hooks/index.js');
require('../../../tokens/index.js');
var popupManager = require('../../../utils/popup-manager.js');
var util = require('../../../utils/util.js');
var index = require('../../../hooks/use-locale/index.js');
var configProvider = require('../../../tokens/config-provider.js');

const configProviderProps = props.buildProps({
  ...index.useLocaleProps,
  size: {
    type: String,
    values: ["large", "default", "small"]
  },
  button: {
    type: props.definePropType(Object)
  },
  zIndex: {
    type: Number
  }
});
var ConfigProvider = vue.defineComponent({
  name: "ElConfigProvider",
  props: configProviderProps,
  setup(props, { slots }) {
    index.provideLocale();
    vue.provide(configProvider.configProviderContextKey, props);
    vue.watch(() => props.zIndex, () => {
      if (util.isNumber(props.zIndex))
        popupManager.PopupManager.globalInitialZIndex = props.zIndex;
    }, { immediate: true });
    return () => {
      var _a;
      return (_a = slots.default) == null ? void 0 : _a.call(slots);
    };
  }
});

exports.configProviderProps = configProviderProps;
exports["default"] = ConfigProvider;
